### ISSUE TITLE
Read submodule defaults from their own config directory (PR welcome)

### ISSUE THREAD

**Issue by lorenwest** [node-config/node-config#225]
Proposal to add 1-arg behavior to `config.setModuleDefaults(moduleName, defaults)`.  If the 2nd argument isn't specified, it loads defaults from the /config directory within the module specified by `moduleName`.

This serves many purposes:

```
* It moves submodule configurations out of code
* It puts submodule configuration into a well known place
* It provides a common place to put configuration documentation
* It allows running the submodule as an app, with overrides for tests, etc.
* It prevents namespace clash by utilizing the submodule package name
* It doesn't break backwards compatibility
* The 2-arg mechanism can be deprecated, with ample warning
```

When this change is made, change the wiki page to document only this behavior.  Make a note about the features the 2-arg behavior doesn't have, and that it will be going away in the future (at least 1 year).

Establish a convention for submodules that they be configured by passing overrides to the module constructor, or by overrides in the app /config directory.  This allows the submodule to be used by apps that also use `node-config` as well as apps that use other configuration systems.

**Comment by lorenwest** [comment]
This will still have the problem outlined by issue #201 so maybe some progress on that issue can be made at the same time.

**Comment by lorenwest** [comment]
The new `setModuleDefaults()` will need to specify the module, since `node-config` isn't able to detect the module directory of the code calling `setModuleDefaults()`.  

Existing method:

    setModuleDefaults(moduleName, configObject);

Proposed alternative:

    setModuleDefaults(module);

If a module is passed as the first arg, the module name will be retrieved from the module's package.json file, and the default config will be loaded from the /config/default.{type} file under the module directory.

**Comment by Bruce17** [comment]
I forked your module and implemented a working fix to this issue (I needed a solution asap).

Have a look at it: https://github.com/Bruce17/node-config/commit/9a6b19c5df601580dd2aafa354c45b7723626d1e

I implemented a new method called `initConfig` which accepts a directory. After loading the module `config`, I call this method to re-init the config module with a additional directory.

```js
// Allow the configs to be overwritten
process.env.ALLOW_CONFIG_MUTATIONS = true;

// Force the module "node-config" to load this module's config files
var config = require('config');
config.initConfig(__dirname + '/../config/');
```

This could be a way to solve this issue. I'm currently using this solution in a working project.

**Comment by lorenwest** [comment]
Your solution is more of a generic "add configurations from any directory", and this issue is more "give submodules a consistent place to store their defaults".  The latter can be implemented using the former, but I'm not a fan of having configurations spread all over the place, so I'm less likely to want to expose a generic ability to do this.

**Comment by tommedema** [comment]
Still having this issue. For now the only solution seems to be a workaround like this

```
function loadConfig(configFolder) {
    process.env.SUPPRESS_NO_CONFIG_WARNING = 'y';
    delete require.cache[require.resolve('config')];
    var oldConfigFolder = process.env.NODE_CONFIG_DIR;
    process.env.NODE_CONFIG_DIR = configFolder;
    var config = require('config');
    process.env.NODE_CONFIG_DIR = oldConfigFolder;
    return config;
}
```

**Comment by alexindigo** [comment]
@lorenwest At first I was thinking on a PR, but then I realized that proposed change was too big and to radical, so there is a light version of `node-config` now, that addresses all the issues I found here, that incompatible with the node-config's nature – http://npmjs.org/configly

It supposed to be backwards-compatible, excepts the parts that made node-config hard to implement those requests (e.g. read/write of `NODE_CONFIG` variable).

**Comment by iMoses** [comment]
Maybe this can be approached in a method similar to `jQuery.noConflict()` method.
`node-config` is sort of a singleton, cached and managing a single instance of `Config`.
My proposal it to provide a method that initiates a different instance of `Config` and returns it to the user.
As it's not suppose to be a breaking change in any way, we can also provide an `options` argument to override any `ENV_VAR` that should never be override by sub-modules.

```
const config = require('config').createModuleInstance({
  configDir: '...',
  appInstance: '...',
});
```

**Comment by markstos** [comment]
So the way I would see this playing out is for a custom project, you would continue to use `node-config` using the current behavior.  If you publishing a module on a `npm` and wanted to use `node-config` privately (internally), you would use the proposed behavior.

A final case to consider is an `npm` module that exposes the config system and what's users to define their own configuration in the parent project.  I guess that would also be supported by using the current `node-config` behavior, which would look for a single configuration relative to $CWD or $NODE_CONFIG_DIR.

This propose looks sound to me, although I dislike `util` name spaces in general.

**Comment by iMoses** [comment]
@markstos that's exactly why I hinted on version 4.
I really hope to force the use of `get()`, remove the possibility of tempering with the immutable configurations, remove all reserved words and add more relevant functionality on top of the `config` instance :) 

@lorenwest I really wanna help around if you'll let me, I'm with you on keeping it simple and without dependencies

**Comment by iMoses** [comment]
> So the way I would see this playing out is for a custom project, you would continue to use `node-config` using the current behavior. If you publishing a module on a `npm` and wanted to use `node-config` privately (internally), you would use the proposed behavior.
> 
> A final case to consider is an `npm` module that exposes the config system and what's users to define their own configuration in the parent project. I guess that would also be supported by using the current `node-config` behavior, which would look for a single configuration relative to $CWD or $NODE_CONFIG_DIR.

Everything you said is correct and yet I'm withdrawing my proposal.

I've been thinking about the many issues opened about this problem and figured it's not the optimal solution. The main problem users the getting is that in case you run inside a sub-package or a cli application you don't necessarily want to use env-vars to avoid stepping on your users toes. Our problem is that for us it's a feature, having an already set config instance when you first import the library.

My new solution is to separate between the code with defines the `Config` class and the code which initializes it. We create a new initialization method that will be used in the main file the this package, populating all the initialization options according to today's initialization logic (`configDit: process.env.CONFIG_DIR || process.env.NODE_CONFIG_DIR,` etc).

This way if you wanna load `node-config` in a sub-package and not interfere with the main package, you can simply call the initialization method yourself without your own options from a different entry-point, and avoid triggering the main instance at all.

```
const initConfig = require('config/module');  // temp names
const config = initConfig({
  configDir: __dirname + '/config',
  appInstance: process.env.NODE_APP_INSTANCE,
});

```

**Comment by markstos** [comment]
@iMoses The new proposal needs to address how it would work in a "sub-module" that has several modules of it's own.  One of the conveniences of `config` is that you can just use `require('config')` in every  file in the project, which is clean and simple.

A sub project using this approach could still provide a clean import statement for most files by creating a tiny `myconfig` class, which would include the explicit initialization of `config` shown in your example. 

The rest of the project could then do just `const config = require('myconfig')`.

**Comment by iMoses** [comment]
@markstos That's how I see it.. Sounds good? 
I didn't get from the comment if you're in favor or just laying down the process :)
We can think about improving it further if necessary, though it will most likely increase the complexity of the solution, so I wonder if it's necessary.

**Comment by markstos** [comment]
@iMoses I'm in favor. I think it's OK if NPM authors have to do a little more work. They are minority of users.

**Comment by lorenwest** [comment]
Our company operates with [inner source](https://en.wikipedia.org/wiki/Inner_source) practices, and as such, teams manage their own configs in their own modules. So at least in our case this isn't the _minority_.

We've written our own support because we believe doing this well requires adherence to a particular naming convention (module name at the top) and didn't feel that requiring our convention was necessarily appropriate for a generic module.

That said, if we decide to move beyond simple submodule configurations to a more advanced mechanism modeled after NPM modules, I'd like to see a separate discussion of additional capabilities.

For us it's this:

* Know where your configs are - always under /config of a module
* Recognize all modules need to run as both `main` and `sub` module.
* Recognize that dependencies are deep - not all known to the `main` module

Once you open this door you realize that anything static - like listing the config directories in a PATH type variable - won't work. It's like re-living the horror of `$JAVA_PATH`, and the best way to stifle modularity.

In other words, the next advancement should be modeled to work with NPM modules. We don't need any more incremental improvements to the existing implementation.

**Comment by iMoses** [comment]
@lorenwest I'll work on a code proposal to address these issues. Clould take me about two weeks, as I'm going on vacation, but we'll see. I'll have some time to burn on flights :)

Only issue I want to flag out in advance, this is going to be a big architectural change, unlike my previous PRs. I'll try to change as little of the logic itself as possible, so it'd be easier to review, but it will be a breaking change on the way to the next major version.

**Comment by lorenwest** [comment]
I would suggest opening a thread on submodule configuration and come to consensus before coding it up. 

My company has done this on top of `node-config` without introducing breaking architectural changes, so I know it can be done. The issues aren't as much coding as they are coming to consensus and involving a wider audience for input.

**Comment by iMoses** [comment]
@lorenwest Now that I'm back home I'll move things forward in that regard..
Do you think we can also start a new thread discussing the next major release, or do you prefer to avoid this issue at this point (until we maxed all our options on v3)?

**Comment by lorenwest** [comment]
I'd suggest starting one thread per feature, and in the discussion we can see if it requires a major version update. If we're creative, it won't require a major version change, and people can enjoy the feature as soon as it rolls out.

I feel it's our responsibility as core module developers to reduce the overall developer burden by making most (if not all) changes backward compatible. The multiplier on our time is huge, and it's worth spending extra time building features that have no downstream cost.

If we're creative we can introduce major philosophical changes without breaking code.

**Comment by iMoses** [comment]
I agree that we should try to avoid breaking changes, but we currently have multiple open issues regarding the config immutability that I don't think we can fix without breaking a thing or two.

That being said, I think that we can avoid breaking anything for users who started with `^2.0` and never accessed the object directly (only thoguh `get`) as recommended by the documentation. It would still be considered a breaking change, as we'll loose backwards compitability with version 1, but I assume most users would have a seamless migration process.

That's basically what I wanted to discuss in a new thread if you think it deserves one.. I do

**Comment by lorenwest** [comment]
We can start (or continue) threads around independent features, and see where we end up with major version updates. 

I'm not a fan of starting a 4.x thread, and having it be the dumping grounds for breaking features.

**Comment by jdmarshall** [comment]
I believe that the Load object introduced in 4.0 satisfies the request made here. No eggs were broken.

However a *lot* of functions were marked deprecated in the process.
